backend/plugin/db/oracle/oracle.go

```diff
@@ -167,16 +171,65 @@ func (driver *Driver) QueryConn(ctx context.Context, conn *sql.Conn, statement s
 	return results, nil
 }
 
-func getOracleStatementWithResultLimit(stmt string, limit int) string {
-	return fmt.Sprintf("SELECT * FROM (%s) WHERE ROWNUM <= %d", stmt, limit)
+func (driver *Driver) getVersion(ctx context.Context) (string, error) {
+	var fullVersion string
+	query := "SELECT BANNER FROM v$version WHERE BANNER LIKE 'Oracle%'"
+	if err := driver.db.QueryRowContext(ctx, query).Scan(&fullVersion); err != nil {
+		if err == sql.ErrNoRows {
+			return "", common.FormatDBErrorEmptyRowWithQuery(query)
+		}
+		return "", util.FormatErrorWithQuery(err, query)
+	}
+
+	var version string
+	tokens := strings.Fields(fullVersion)
+	for _, token := range tokens {
+		if semVersionRegex.MatchString(token) {
+			version = token
+			continue
+		}
+	}
+
+	return version, nil
+}
+
+func (driver *Driver) getOracleStatementWithResultLimit(ctx context.Context, stmt string, limit int) (string, error) {
+	if driver.dbVersion == 0 {
+		version, err := driver.getVersion(ctx)
+		if err != nil {
+			return "", err
+		}
+		versionIdx := strings.Index(version, ".")
+		if versionIdx < 0 {
+			return "", errors.New("instance version number is invalid")
+		}
+		driver.dbVersion, err = strconv.Atoi(version[:versionIdx])
+		if err != nil {
+			return "", err
+		}
+	}
```

Comment:
When we get errors here, we'd better log them and use the default option to allow queries. In other words, use the failed-open strategy.